design-proposals/klip-46-date-and-time-data-type-support.md

```diff
@@ -0,0 +1,297 @@
+* `UNIX_TIME(time)` -  returns an INTEGER number of milliseconds that have passed 00:00:00.000 and the specified time
+* `CONVERT_TZ(time, from_tz, to_tz)` - converts a time from one timezone to another
+* `DATEADD(time unit, integer, date)` - Adds an interval to the date. The time unit must be `DAYS` or `YEARS`. If it is not,
+then the function will throw an error
+* `DATESUB(time unit, integer, date)` - Subtracts an interval from the date. The time unit must be `DAYS` or `YEARS`. If it is not,
+then the function will throw an error
+* `TIMEADD(time unit, integer, time)` - Adds an interval to the time. If the result falls outside of
+the `00:00:00.000` to `23:59:59.999` range, then the result will be adjusted.
+* `TIMESUB(time unit, integer, time)` - Subtracts an interval from the time. If the result falls outside of
+the `00:00:00.000` to `23:59:59.999` range, then the result will be adjusted.
+
+## Design
+### Serialization/Deserialization
+
+The DATE type will be handled by the `java.sql.Date` class within KSQL and the TIME type will be
+handled by `java.sql.Time`. The corresponding Kafka Connect types are [org.apache.kafka.connect.data.Date](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Date.html)
+and [org.apache.kafka.connect.data.Time](https://kafka.apache.org/0100/javadoc/org/apache/kafka/connect/data/Time.html).
+Both are represented as integers in Schema Registry with tags to identify their logical types.
+
+#### Avro
+Avro schemas represent dates as
+
+```json
+{
+  "type": "int",
+  "logicalType": "date"
+}
+```
+
+And time as
+```json
+{
+  "type": "int",
+  "logicalType": "time-millis"
+}
+```
+
+The Avro deserializer ksqlDB uses support these.
+
+#### JSON/Delimited
+
+Dates will get stored in JSON and CSV files as an int number of days since Unix Epoch. Times will
+get stored as an int number of milliseconds since 00:00:00.000
+
+The KSQL JSON and delimited deserializers will be updated to parse dates and times.
+
+#### Protobuf
+Protobuf has a `google.type.TimeOfDay` type for time and a `google.type.Date` type for dates.
+The Protouf deserializer KSQL uses supports this.
+
+### Casting
+#### TIMESTAMP as DATE
+Casting from TIMESTAMP to DATE discards the time portion of the TIMESTAMP:
+```roomsql
+-- ts is the timestamp, '2007-01-01T03:20:45'
+SELECT CAST(ts AS DATE) FROM s;
+> 2007-01-1
+```
+#### TIMESTAMP as TIME
+Casting from TIMESTAMP to DATE discards the date portion of the TIMESTAMP:
+```roomsql
+-- ts is the timestamp, '2007-01-01T03:20:45'
+SELECT CAST(ts AS TIME) FROM s;
+> 03:20:45
+```
+#### DATE AS TIMESTAMP
+Casting from DATE to TIMESTAMP adds `00:00:00.000` as the time portion to the DATE:
+```roomsql
+-- d is the date, '2007-01-01'
+SELECT CAST(d AS TIMESTAMP) FROM s;
+> 2007-01-1T00:00:00
+```
+#### TIME AS TIMESTAMP
+Casting from TIME to TIMESTAMP adds `1970-01-01` as the date portion to the TIME:
```

Comment:
I don't think we even need to support this. It wouldn't add value IMO, and would just create surface area for surprising behavior and/or silent mistakes.

Reply:
Yeah that makes sense. Removed.